Snowflake.Data/Core/TomlConnectionBuilder.cs

```diff
@@ -0,0 +1,152 @@
+using Tomlyn;
+using Tomlyn.Model;
+
+namespace Snowflake.Data.Core
+{
+    internal class TomlConnectionBuilder
+    {
+        private const string DefaultConnectionName = "default";
+        private const string DefaultSnowflakeFolder = ".snowflake";
+        private const string DefaultTokenPath = "/snowflake/session/token";
+
+        internal const string SnowflakeDefaultConnectionName = "SNOWFLAKE_DEFAULT_CONNECTION_NAME";
+        internal const string SnowflakeHome = "SNOWFLAKE_HOME";
+
+        private static readonly SFLogger s_logger = SFLoggerFactory.GetLogger<SnowflakeDbConnection>();
+
+        private readonly Dictionary<string, string> _tomlToNetPropertiesMapper = new Dictionary<string, string>(StringComparer.InvariantCultureIgnoreCase)
+        {
+            { "DATABASE", "DB" }
+        };
+
+        private readonly FileOperations _fileOperations;
+        private readonly EnvironmentOperations _environmentOperations;
+
+        public static readonly TomlConnectionBuilder Instance = new TomlConnectionBuilder();
+
+        private TomlConnectionBuilder() : this(FileOperations.Instance, EnvironmentOperations.Instance)
+        {
+        }
+
+        internal TomlConnectionBuilder(FileOperations fileOperations, EnvironmentOperations environmentOperations)
+        {
+            _fileOperations = fileOperations;
+            _environmentOperations = environmentOperations;
+        }
+
+        public string GetConnectionStringFromToml(string connectionName = null)
+        {
+            var tomlPath = ResolveConnectionTomlFile();
+            var connectionToml = GetTomlTableFromConfig(tomlPath, connectionName);
+            s_logger.Debug($"Reading connection parameters from file using key: {connectionName} and path: {tomlPath}");
+            return connectionToml == null ? string.Empty : GetConnectionStringFromTomlTable(connectionToml);
+        }
+
+        private string GetConnectionStringFromTomlTable(TomlTable connectionToml)
+        {
+            var connectionStringBuilder = new StringBuilder();
+            var tokenFilePathValue = string.Empty;
+            var isOauth = connectionToml.TryGetValue("authenticator", out var authenticator) && authenticator.ToString().Equals("oauth");
+            foreach (var property in connectionToml.Keys)
+            {
+                var propertyValue = (string)connectionToml[property];
+                if (isOauth && property.Equals("token_file_path", StringComparison.InvariantCultureIgnoreCase))
+                {
+                    tokenFilePathValue = propertyValue;
+                    continue;
+                }
+                var mappedProperty = _tomlToNetPropertiesMapper.TryGetValue(property, out var mapped) ? mapped : property;
+                connectionStringBuilder.Append($"{mappedProperty}={propertyValue};");
+            }
+
+            AppendTokenFromFileIfNotGivenExplicitly(connectionToml, isOauth, connectionStringBuilder, tokenFilePathValue);");
```

Comment:
You have unexpected "); here and it doesn't allow to build the code.

Reply:
Fixed thanks!!!